### ISSUE TITLE
Add setting to disallow "arrow function declarations" at the global/top level

### ISSUE THREAD

**Issue by jfroelich** [getify/eslint-plugin-proper-arrows#34]
I would like to forbid top level function declarations that use arrow syntax. This works using global.

```js
const bad = () => {
  console.log('bad');
};
```

But I would like to allow top level arrow functions present within a function call. I cannot get this to work because these are considered global. I recognize they are indeed global. I would like to carve out an exception here. This issue here blocks me from being able to benefit from this plugin and I am going to have to disable it.

```js
const things = [].map(x => console.log('eh, this is ok, it is in a function call');
```

**Comment by getify** [comment]
Can you explain why you want to carve out an exception? What makes a global arrow in the argument position better/more acceptable than a global arrow in other positions (like var or prop assignments)?

**Comment by jfroelich** [comment]
I am trying to forbid the use of top level arrow function syntax for function declaration. `func-style` is similarly insufficient.

I am trying to get a team of developers to all write functions using a conventional style `function functionName() {}` and stop using the `const functionName = () {...}` style. This is a trivialized example. The assignment style has led to some really awkward looking code and I just want the team to move away from it. I control the keys to the linter so I really want to enforce this.

However, we have a great deal of code, such as every single script file, that makes calls like `main().catch(error => console.error(error));`. Again, a trivialized example, it gets a lot worse. Every single one of these scripts now has a linting error. I would like to continue to allow these arrow functions when used in this way, while at the same time, disallow use of them for function declarations. Your plugin is the closest thing I found. It is just missing the ability to make this one case an exception.

**Comment by getify** [comment]
Appreciate the additional detail. I'm open to exploring adding this extra "exception" mode, but I always like to deeply understand a problem and explore any possible alternatives before committing to extending the code.

I know you have a variety of these "global arrow functions" you're concerned about. Are they *always* in an inline callback-argument position?

IOW, do you have things like this that you also want to allow:

```js
someFunction({
   whatever: true,
   count: 12,
   onError: err => console.log(err)
});
```

That style is crops up quite common when you move from "positional arguments" to "named arguments" style coding, where function calls all now have a single object with properties as "named arguments" (fairly common idiom these days).

Or what about stuff like this:

```js
someFunction.apply(null,[ 42, "hello", err => console.log(err) ]);
```

Actually, more often, the calls happen where an array is setup separately (and includes an arrow function, and then it's just applied, like:

```js
var args = [ 42, "hello", err => console.log(err) ];

// later:
someFunction.apply(null,args);

// or:
someFunction(...args):
```

Do you do have any "IIAFEs", like this?

```js
(() => {
   // ..
})();

// or

var something = (()=>{ 
   // ..
   return 42;
})();
```

Or these, which are rarer and more contrived, but still legal and possible:

```js
var str = someSpecialTagFn`The text I want to show you is: ${ () => whatever().toUpperCase() }!!`;
```

... where the `someSpecialTagFn` tag-function accepts an arrow function as an "argument" via a template-string expression.

My point is, if we are to discuss the "scope" of this exception mode, we need to understand just how broad it needs to be. And we need to balance the decision making, in that if we make the exception mode narrow (say, "arguments only") now, to suit your specific needs, what are the chances later that someone else (or you, 6 months from now!) comes along and wants to be able to grant exceptions for more of these other variations?

----

> `main().catch(error => console.error(error));`

That's definitely a common pattern. I'm not sure how many of your linting errors are from stuff like that, but... one option to consider (which I think is better even regardless of linter styles) is what's called "point-free" in the FP world:

```js
main().catch(console.error)
```

That's functionally equivalent when the outer function expression (the arrow) has the same, or compatible, signature as the inner function call -- you can simply flatten it one level.

Same goes for simple callbacks provided to things like `map(..)`:

```js
[ 1, 2, 3 ].map(v => lookupID(v));

// can be:

[ 1, 2, 3 ].map(lookupID);
```

And in cases where you are doing the same sort of inline function expression a bunch, like if this arrow appears in a lot of places:

```js
var strs = [ "some", "Strings", "OF", "value" ].map(v => v.toUpperCase());
```

It might be a "better fix" instead of allowing that exception to just define a single re-usable function like:

```js
function upperString(v) { return v.toUpperCase(); }

// and then using:
var strs = [ "some", "Strings", "OF", "value" ].map(upperString);
```

In other words, I have found that in a LOT of cases, the need for these inline arrow function expressions can be reduced, or eliminated entirely, by taking a step back and trying a few simple refactorings.

**Comment by jfroelich** [comment]
> Appreciate the additional detail. I'm open to exploring adding this extra "exception" mode, but I always like to deeply understand a problem and explore any possible alternatives before committing to extending the code.

Thank you for looking into it.

> I know you have a variety of these "global arrow functions" you're concerned about. Are they _always_ in an inline callback-argument position?

Not really. I would say that I am trying to distinguish the one situation, the declaration of a function at module scope (e.g. not a "lambda" in the C++ sense), from all other situations whatever the appropriate names for those other situations are (expression-based?).

> IOW, do you have things like this that you also want to allow:
> 
> ```js
> someFunction({
>    whatever: true,
>    count: 12,
>    onError: err => console.log(err)
> });
> ```

I would prefer that this still be allowed.

> That style crops up quite commonly when you move from "positional arguments" to "named arguments" style coding, where function calls all now have a single object with properties as "named arguments" (fairly common idiom these days).

Yes. We have a very large amount of code that uses "named" function arguments. I will point out an aside here that we are simultaneously attempting to minimize the amount of logic that is expressed within object literals. The problem with named arguments is that the object literal you express as the parameter turns into a megabyte sized JSON looking thing of its own, and starts to get difficult to parse. Especially with lambdas. So even though we use named arguments, we discourage defining functions in named arguments. But that discouragement does not amount to a linting error.

> Or what about stuff like this:
> 
> ```js
> someFunction.apply(null,[ 42, "hello", err => console.log(err) ]);
> ```

I would prefer this continue to be allowed. This is an expression use case, not the declaration of a new function, from my POV.

> Actually, more often, the calls happen where an array is setup separately (and includes an arrow function), and then it's just applied later, like:
> 
> ```js
> var args = [ 42, "hello", err => console.log(err) ];
> 
> // later:
> someFunction.apply(null,args);
> 
> // or:
> someFunction(...args):
> ```

I would prefer this be allowed. This is expressing an arrow function as an expression or callback param, not declaring a new top level named function.

> Do you do have any "IIAFEs", like this?
> 
> ```js
> (() => {
>    // ..
> })();
> 
> // or
> 
> var something = (()=>{ 
>    // ..
>    return 42;
> })();
> ```

Yes, we have a very large number of IIFEs. Especially for older code that has not been modernized. We are actually in the process of moving everything to typescript. At the same time, we are moving away from all IIFE use. I am trying to get everyone to use ES6 module import style. We want plain imports and exports. With modules, the need for an IIFE is gone. Removing IIFEs also makes the code more readable. 

What we are basically doing is converting every IIFE into the equivalent

```
async function main() {
}

main().catch(console.error);
```

> Or these, which are rarer and more contrived, but still legal and possible:
> 
> ```js
> var str = someSpecialTagFn`The text I want to show you is: ${ () => whatever().toUpperCase() }!!`;
> ```
> 
> ... where the `someSpecialTagFn` tag-function accepts an arrow function as an "argument" via a template-string expression.

I am currently focused on trying to have no functions defined in template literals. We have a bunch of this, and it has led to writing entire programs inside of template literals, and it is not a pretty sight. While not exactly a linting error, it is something I discourage. I would prefer this to not be a linting error in the context of this github issue. It would be a different discussion. Perhaps some other kind of rule.

> My point is, if we are to discuss the "scope" of this exception mode, we need to understand just how broad it needs to be. And we need to balance the decision making, in that if we make the exception mode narrow (say, "arguments only") now, to suit your specific needs, what are the chances later that someone else (or you, 6 months from now!) comes along and wants to be able to grant exceptions for more of these other variations?

Great point. I would like to try and narrow the request to specifically the case of a module-scope named function definition, as opposed to a module-scope function expression.

> > `main().catch(error => console.error(error));`
> 
> That's definitely a common pattern. I'm not sure how many of your linting errors are from stuff like that, but... one option to consider (which I think is better even regardless of linter styles) is what's called "point-free" in the FP world:
> 
> ```js
> main().catch(console.error)
> ```

Yes. In fact most of the code was originally like what you suggest here. However, there is a strange unrelated typescript linting error related to "this" binding that forbids your suggestion. Replacing `console.error` with `error => console.error(error)` is just a quick way of avoiding that linting error. That is another annoyance that I hope to deal with another day and probably outside the scope of this.

> That's functionally equivalent when the outer function expression (the arrow) has the same, or compatible, signature as the inner function call -- you can simply flatten it one level.
> 
> Same goes for simple callbacks provided to things like `map(..)`:
> 
> ```js
> [ 1, 2, 3 ].map(v => lookupID(v));
> 
> // can be:
> 
> [ 1, 2, 3 ].map(lookupID);
> ```

Yes, up until we run into the this binding error.

> 
> And in cases where you are doing the same sort of inline function expression a bunch, like if this arrow appears in a lot of places:
> 
> ```js
> var strs = [ "some", "Strings", "OF", "value" ].map(v => v.toUpperCase());
> ```

So far I have minimized the use of any higher order like logic.

> It might be a "better fix" instead of allowing that exception to just define a single re-usable function like:
> 
> ```js
> function upperString(v) { return v.toUpperCase(); }
> 
> // and then using:
> var strs = [ "some", "Strings", "OF", "value" ].map(upperString);
> ```

Certainly.

> In other words, I have found that in a LOT of cases, the need for these inline arrow function expressions can be reduced, or eliminated entirely, by taking a step back and trying a few simple refactorings.

Yep. Trying to use the linter to encourage transitioning. There are over 100 repos, and there are 50 reasons to refactor each one. This is just one step. Trying to nab this one situation where we have inconsistent ways of defining functions.

My dream would be that all modules take the form of something like this:

```typescript
async function a() {
}

function b() {
}

async function c() {
   const someLambdaLocalToC = () => 1;
}
```

Just a bunch of top level separate units that focus on definition, not statement evaluation order. That is what I am going for.

**Comment by getify** [comment]
Your answers are super helpful. That's why I like to explore before jumping right in! :)

I thought you just wanted a narrow exception for arrows as *arguments*. But it sounds like what you actually want is a mode like "global" but that *only* disallows arrow "declarations" (that is, assignments of an arrow function to a variable), and otherwise allow all other forms of global arrow expressions. It could be called "global-declarations" or something like that.

Does that sound like a reasonable summary of that request?

Assuming that's correct, let me further clarify, then...

```js
function doSomething() {
    const handleResults = (result) => render(result.rows);
    anotherThing(42,handleResults);
}
```

Should that be allowed? IOW, is it *only* top-level "arrow function declarations" that you want to disallow, or all such arrow function declarations at any level?

If that snippet is OK, I guess I'd like to hear why arrow function declarations inside other functions is OK, in your opinion, but not OK at the global level. What "badness" are you really trying to avoid? Is it *just* stylistic preference, or is there a deeper problem you're hoping to avoid, that's more present at the global level and not present inside scopes?

OTOH, if that snippet above would be discouraged, would this version be OK?

```js
function doSomething() {
    anotherThing(42,(result) => render(result.rows));
}
```

So all we did here to "fix" was just inline the arrow function. Is that actually "fixed" acceptably? Why?

Again, just hoping to get more insight into your thinking, so we design a feature that's well aligned and suitable.

**Comment by jfroelich** [comment]
> Your answers are super helpful. That's why I like to explore before jumping right in! :)
> 
> I thought you just wanted a narrow exception for arrows as _arguments_. But it sounds like what you actually want is a mode like "global" but that _only_ disallows arrow "declarations" (that is, assignments of an arrow function to a variable), and otherwise allow all other forms of global arrow expressions. Such a mode could be called "global-declarations" or something like that.
> 
> Does that sound like a reasonable summary of your request?

Yes.

> 
> Assuming that's correct, let me further clarify, then...
> 
> ```js
> function doSomething() {
>     const handleResults = (result) => render(result.rows);
>     anotherThing(42,handleResults);
> }
> ```
> 
> Should that be allowed? IOW, is it _only_ top-level "arrow function declarations" that you want to disallow, or all such arrow function declarations at any level?

This should be allowed. I only want to disable top level arrow function declarations. I want to continue to allow arrow function declarations anywhere other than the top level.

> If that snippet is OK, I guess I'd like to hear why arrow function declarations inside other functions is OK, in your opinion, but not OK at the global level. What "badness" are you really trying to avoid? Is it _just_ stylistic preference, or is there a deeper problem you're hoping to avoid, that's more present at the global level and not present inside scopes?

Stylistic. I am focused on readability, uniformity, and ease of maintenance. I distinguish between "lambdas" that are local helper functions to a function, scoped to that function, and top level functions. I am trying to minimize the amount of nesting because the amount of nesting borders on callback hell at some point. Same goes with use of fluent apis that chain a bunch of methods. The code starts to get pretty unreadable. I suppose a similarity might be the preference for using `for..of` over `forEach`. Arrow functions were a cool addition, but the javascript community has gone to far, and started using them anywhere. They were, in my opinion, only intended to help write more concise lambdas and avoid the `var that = this` stuff.

Terser syntax is nice and all but it has a limit after which it harms code readability. Lambdas local to function bodies are within the limit. Top level lambdas are just an alternate way to express a conventional function declaration and should be forbidden.

> OTOH, if that snippet above would be discouraged, would this version be OK?
> 
> ```js
> function doSomething() {
>     anotherThing(42,(result) => render(result.rows));
> }
> ```
> 
> So all we did here to "fix" was just inline the arrow function. Is that actually "fixed" acceptably? Why?

That is fixed acceptably. From an opinionated point of view, however, once you start down this path, you end with code doing `a(b(c(d(e))))` kind of stuff. At some point you cross over a threshold where the code transitions from easy to read to a mess. At some point you are writing lisp/scheme/haskell and are not writing c-like imperative code and you lose a lot of that simplicity. That said, we do both your first example and the second all the time, and both are fine.

> Again, just hoping to get more insight into your thinking, so we design a feature that's well aligned and suitable.

Thank you.

**Comment by getify** [comment]
One last clarification (I hope!):

```js
var {
   cb: myFunc = (evt) => evt.preventDefault()
} = whatever();
```

There, the "declaration" of the global arrow is happening as part of the destructuring, in the default assignment clause. Is this a "declaration" you'd expect the rule to block, or are focused only on these:

```js
var f = () => ..
let g = () => ..
const g = () => ..
```

**Comment by jfroelich** [comment]
> ```js
> var {
>    cb: myFunc = (evt) => evt.preventDefault()
> } = whatever();
> ```

This would be allowed since from my POV this function is just another expression, not a top level declaration.

> ```js
> var f = () => ..
> let g = () => ..
> const g = () => ..
> ```

Yes, only focused on these 3 exact cases at the top level.

**Comment by jfroelich** [comment]
> ```js
> var cb;
> if (isSomething) {
>    cb = () => g(2);
> }
> else {
>    cb = () => h(3);
> }
> ```

This one is tougher than I thought. I do not have a straightforward answer here with any type of clear logic at first glance.

I think, subjectively, I lean towards saying this is allowed. In some sense this is just deferred assignment of a function expression.

**Comment by getify** [comment]
Similar question about `var cb = isSomething ? () => g(2) : () => h(3);`

**Comment by jfroelich** [comment]
> Similar question about `var cb = isSomething ? () => g(2) : () => h(3);`

Allowed.

**Comment by getify** [comment]
So the narrowest definition for this new mode (for disallowance) is:

any top-level scope assignment that's directly part of a declaration (`var`, `let`, or `const`) where the value being assigned is exactly and only an arrow function.

Sound reasonable?

**Comment by jfroelich** [comment]
Yes. Thank you for narrowing the scope of the request so well. Tough to envision all the situations.

**Comment by getify** [comment]
I had landed this feature in v11.0.0.

Here's the documentation for the rule mode `"global-declaration"`: https://github.com/getify/eslint-plugin-proper-arrows#rule-where-configuration-global-declaration

To accomplish what you were hoping for in this thread, you need to explicitly set `"global-declaration"` to `false`, while either turning `"global"` on (`true`) explicitly, or letting it just be on by default.

**NOTE:** you can *also* do the opposite, which is to error *only* on these global arrow declarations and not on any other global arrow expressions, by doing the reverse: setting `"global"` explicitly to `false`, and setting `"global-declaration"` explicitly to `true`. That may be useful in debugging processes, in terms of you verifying this rule mode is working the way you hope. Please check it out and let me know!

**Comment by jfroelich** [comment]
You ROCK!

**Comment by jfroelich** [comment]
This worked exactly how I wanted it to:

```json
{
    "@getify/proper-arrows/where": [
      "error",
      {
        "global": false,
        "global-declaration": true,
        "property": false,
        "export": false,
        "trivial": false
      }
    ]
}
```

This is an error:

```typescript
// arrow function declaration
export const topLevelArrowBad = (test: string) => {
  console.log(test);
};
```

This is not an error:

```typescript
// arrow function
process.on('unhandledRejection', error => {
  console.log(error);
});
```

This is not an error:

```js
async function main() {
  console.log('main');
}

main().catch(console.error);
```

Thanks!

**Comment by jfroelich** [comment]
Sorry for the spam. One minor note. Using the property name 'global-declaration' instead of 'globalDeclaration' threw me off for a second, just because of the convention established by other eslint rules. I do not personally care about this, but your other users might. Thanks!

**Comment by getify** [comment]
I debated on what to call it... maybe I made the wrong call here, but I felt like camel-casing would make things more confusing.
